source/interfaces/event_queue.c

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "jd_protocol.h"
+
+typedef struct {
+    uint32_t timestamp;
+    uint8_t service_size;
+    uint8_t service_number;
+    uint16_t service_command;
+    uint8_t data[0];
+} ev_t;
+
+struct event_info {
+    ev_t *buffer;
+    uint16_t qptr;
+    uint8_t counter;
+};
+static struct event_info info;
+
+static inline uint32_t ev_size(ev_t *ev) {
+    return sizeof(ev_t) + ((ev->service_size + 3) & ~3);
+}
+
+static inline ev_t *next_ev(ev_t *ev) {
+    return (ev_t *)((uint8_t *)ev + ev_size(ev));
+}
+
+static uint16_t next_event_cmd(uint32_t eventid) {
+    if (eventid >> 24)
+        jd_panic();
+    info.counter++;
+    return JD_CMD_EVENT_MK(info.counter, eventid);
+}
+
+static void ev_init(void) {
+    if (info.buffer)
+        return;
+    info.buffer = jd_alloc(JD_EVENT_QUEUE_SIZE);
+}
+
+static void ev_shift(void) {
+    int words = ev_size(info.buffer) >> 2;
+    uint32_t *dst = &info.buffer->timestamp;
+    uint32_t *src = dst + words;
+    info.qptr -= words << 2;
+    while (words--)
+        *dst++ = *src++;
+}
+
+void jd_process_event_queue(void) {
+    if (info.qptr == 0)
+        return;
+
+    ev_t *ev = info.buffer;
```

Comment:
could be dangerous if info.buffer not init'd?

Reply:
Is initialisation to zero guaranteed? I guess so if it's in globals...

Reply:
added comment; globals are 0-initalized